src/lib/krb5/keytab/kt_file.c

```diff
@@ -462,6 +462,8 @@ krb5_ktfile_start_seq_get(krb5_context context, krb5_keytab id, krb5_kt_cursor *
         /* Wrapped?!  */
         KTITERS(id)--;
         KTUNLOCK(id);
+        free(fileoff);
+        fileoff = NULL;
```

Comment:
Here, `fileoff` has already been set as the value of `*cursorp`; freeing it would leave a dangling pointer in `*cursorp` for the caller.  (Setting `fileoff = NULL` does nothing, as that is a local variable about to go out of scope.)  I'm not sure how this was flagged as a leak, although we do generally try to avoid making the caller free stuff on errors.

Reply:
My bad, here I meant to set `cursorp` to `NULL`, not `fileoff`. It was to avoid returning the address of a memory block that that already freed.

Not freeing `fileoff` here was causing a leak in [`src/kadmin/cli/keytab.c:remove_principal()`](https://github.com/krb5/krb5/blob/krb5-1.21-final/src/kadmin/cli/keytab.c#L451). In case `krb5_kt_start_seq_get()` fails, we leave the function without freeing the memory `cursor` is pointing too.

But since a cursor is just supposed to be a moving pointer to an underlying resource, freeing a cursor is something that should not happen, since it might not be pointing to the beginning of the allocated memory block.

However, I see that it's what is being done in [`krb5_ktfile_end_get()`](https://github.com/krb5/krb5/blob/krb5-1.21-final/src/lib/krb5/keytab/kt_file.c#L512), so maybe I'm just overthinking this.

Reply:
Keytab iteration cursors are allocated objects (not just alias pointers) and do need to be freed.  remove_principal() will still leak the cursor multiple other error cases if krb5_ktfile_start_seq_get() is fixed, so that should be addressed as well.